modules/nf-core/deseq2/differential/templates/deseq_de.r

```diff
@@ -0,0 +1,354 @@
+
+model <- paste(model, opt\$contrast_variable, sep = ' + ')
+
+################################################
+################################################
+## Run DESeq2 processes                       ##
+################################################
+################################################
+
+dds <- DESeqDataSetFromMatrix(
+    countData = round(count.table),
+    colData = sample.sheet,
+    design = as.formula(model)
+)
+
+dds <- DESeq(
+    dds,
+    test = opt\$test,
+    fitType = opt\$fit_type,
+    minReplicatesForReplace = opt\$min_replicates_for_replace,
+    useT = opt\$use_t,
+    sfType = opt\$sf_type
+)
+
+comp.results <-
+    results(
+        dds,
+        lfcThreshold = opt\$lfc_threshold,
+        altHypothesis = opt\$alt_hypothesis,
+        independentFiltering = opt\$independent_filtering,
+        alpha = opt\$alpha,
+        pAdjustMethod = opt\$p_adjust_method,
+        minmu = opt\$minmu,
+        contrast = c(
+            opt\$contrast_variable,
+            c(opt\$treatment_level, opt\$reference_level)
+        )
+    )
+
+################################################
+################################################
+## Generate outputs                           ##
+################################################
+################################################
+
+# Common function to round numerics in data frames - mostly useful for testing,
+# since it doesn't seem to be possible to make outputs entirely reproducible
+# across machienes, even with set.seed(), but the differences are tiny.
+
+prepare_results <- function(x){
+    if (opt\$round_results){
+        format(x, nsmall=8)
+    }else{
+        x
+    }
+}
+
+
+contrast.name <-
+    paste(opt\$treatment_level, opt\$reference_level, sep = "_vs_")
+cat("Saving results for ", contrast.name, " ...\n", sep = "")
+
+# Differential expression table
+
+write.table(
+    prepare_results(data.frame(gene_id = rownames(comp.results), comp.results)),
+    file = 'deseq2.results.tsv',
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+# Dispersion plot
+
+png(
+    'deseq2.dispersion.png',
+    width = 600,
+    height = 600
+)
+plotDispEsts(dds)
+dev.off()
+
+# R object for other processes to use
+
+save(dds, file = 'dds.rld.RData')
+
+# Size factors
+
+sf_df = data.frame(sample = names(sizeFactors(dds)), data.frame(sizeFactors(dds)))
+colnames(sf_df) <- c('sample', 'sizeFactor')
+write.table(
+    sf_df,
+    file = 'deseq2.sizefactors.tsv',
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+# Write specified matrices
+
+if (opt\$write_normalised){
+    write.table(
+        data.frame(gene_id=rownames(counts(dds)), counts(dds, normalized = TRUE)),
+        file = 'normalised_counts.tsv',
+        col.names = TRUE,
+        row.names = FALSE,
+        sep = '\t',
+        quote = FALSE
+    )
+}
+
+if (opt\$write_variance_stabilised){
```

Comment:
I would also just export these two methods as two separate files, not ask.

Reply:
Ok, will do!